apiserver/charms/client.go

```diff
@@ -45,19 +48,22 @@ func (a *API) checkCanRead() error {
 	return nil
 }
 
-// NewAPI returns a new charms API facade.
-func NewAPI(
-	st *state.State,
-	resources facade.Resources,
-	authorizer facade.Authorizer,
-) (*API, error) {
+// NewFacade provides the signature required for facade registration.
+func NewFacade(ctx facade.Context) (*API, error) {
+	var st charmsAccess
+	if ctx.IsCAAS() {
+		st = ctx.CAASState()
+	} else {
+		st = ctx.State()
+	}
 
+	authorizer := ctx.Auth()
 	if !authorizer.AuthClient() {
 		return nil, common.ErrPerm
 	}
 
 	return &API{
-		access:     getState(st),
+		access:     st,
```

Comment:
Urgh. Someone didn't understand Go very well (and a reviewer wasn't paying attention).

I might go and fix this upstream. I see a few other "funny" things in this package.